src/JSInterop/Microsoft.JSInterop/src/DotNetObjectReferenceJsonConverter.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.JSInterop
+{
+    internal sealed class DotNetObjectReferenceJsonConverter<TValue> : JsonConverter<DotNetObjectRef<TValue>> where TValue : class
+    {
+        private static JsonEncodedText DotNetObjectRefKey => DotNetDispatcher.DotNetObjectRefKey;
+
+        public override DotNetObjectRef<TValue> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            long dotNetObjectId = 0;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
```

Comment:
One thing about this - I don't think it would detect an incomplete object. `{ __id: "foo"`

Reply:
The reader should throw at this point - it'll attempt to read till the end of token and complain when it doesn't find a terminator.